go-controller/pkg/clustermanager/routeadvertisements/controller.go

```diff
@@ -0,0 +1,1124 @@
+	}
+
+	// update them
+	hadFRRConfigUpdates, err := c.updateFRRConfigurations(name, frrConfigs)
+	if err != nil {
+		return false, fmt.Errorf("failed updating FRRConfigurations for RouteAdvertisements %q: %w", name, err)
+	}
+
+	// annotate NADs
+	hadNADUpdates, err := c.updateNADs(name, nads)
+	if err != nil {
+		return false, fmt.Errorf("failed annotating NADs for RouteAdvertisements %q: %w", name, err)
+	}
+
+	return hadFRRConfigUpdates || hadNADUpdates, nil
+}
+
+// selectedNetworks is a helper struct that stores information about networks
+// that have been selected by a RouteAdvertisements. It is important that prefix
+// lists are ordered to generate consistent FRRConfigurations.
+type selectedNetworks struct {
+	// networks is an ordered list of selected network names
+	networks []string
+	// vrfs is an ordered list of selected networks VRF's
+	vrfs []string
+	// networkVRFs is a mapping of VRF to corresponding network
+	networkVRFs map[string]string
+	// subnets is an ordered list of all selected network subnets
+	subnets []string
+	// hostSubnets is an ordered list of all selected network subnets specific to a node
+	hostSubnets []string
+	// networkSubnets is a map of selected network names to their ordered network subnets
+	networkSubnets map[string][]string
+	// hostNetworkSubnets is a map of selected network names to their ordered network subnets specific for a node
+	hostNetworkSubnets map[string][]string
+	// prefixLength is a map of selected network to their prefix length
+	prefixLength map[string]uint32
+}
+
+// generateFRRConfigurations generates FRRConfigurations for the route
+// advertisements. Also returns the selected network NADs.
+func (c *Controller) generateFRRConfigurations(ra *ratypes.RouteAdvertisements) ([]*frrtypes.FRRConfiguration, []*nadtypes.NetworkAttachmentDefinition, error) {
+	if ra == nil {
+		return nil, nil, nil
+	}
+
+	// if we are matching on the well known default network label, create an
+	// internal nad for it if it doesn't exist
+	if matchesDefaultNetworkLabel(ra.Spec.NetworkSelector) {
+		_, err := c.getOrCreateDefaultNetworkNAD()
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get/create default network NAD: %w", err)
+		}
+	}
+
+	// gather selected networks
+	var nads []*nadtypes.NetworkAttachmentDefinition
+	nadSelector, err := metav1.LabelSelectorAsSelector(&ra.Spec.NetworkSelector)
+	if err != nil {
+		return nil, nil, err
+	}
+	nads, err = c.nadLister.List(nadSelector)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// validate and gather information about the networks
+	networkSet := sets.New[string]()
+	selectedNetworks := &selectedNetworks{
+		networkVRFs:    map[string]string{},
+		networkSubnets: map[string][]string{},
+		prefixLength:   map[string]uint32{},
+	}
+	for _, nad := range nads {
+		networkName := util.GetAnnotatedNetworkName(nad)
+		network := c.nm.GetNetwork(networkName)
+		if network == nil {
+			// network not yet known by network manager, skip
```

Comment:
oh, so how do we trigger a reconcile again once nm has seen this network? Is it because the NAD gets updated with network id or something?